ts/nni_manager/common/gpu_scheduler/scheduler.ts

```diff
@@ -0,0 +1,246 @@
+     *  Initialize the scheduler.
+     *
+     *  If there is no GPU found, throw error.
+     **/
+    public async init(): Promise<void> {
+        const info = await collectGpuInfo(true);
+        if (info === null) {
+            throw new Error('GpuScheduler: Failed to collect GPU info');
+        }
+        if (info.gpuNumber === 0) {
+            throw new Error('GpuScheduler: No GPU found');
+        }
+
+        for (let i = 0; i < info.gpuNumber; i++) {
+            this.gpus.push({
+                index: i,
+                util: 0,
+                coreUtil: 0,
+                memUtil: 0,
+                active: false,
+                computeActive: false,
+            });
+        }
+
+        this.updateGpus(info);
+    }
+
+    /**
+     *  Update GPUs' utilization info.
+     *
+     *  If `force` is not true, it may use cached result.
+     *
+     *  This is normally unnecessary because it will implicitly update before scheduling.
+     **/
+    public async update(force?: boolean): Promise<void> {
+        const info = await collectGpuInfo(force);
+        if (info === null) {
+            if (force) {
+                throw new Error('GpuScheduler: Failed to update GPU info');
+            }
+            return;
+        }
+        if (info.gpuNumber !== this.gpus.length) {
+            throw new Error(`GpuScheduler: GPU number changed from ${this.gpus.length} to ${info.gpuNumber}`);
+        }
+
+        this.updateGpus(info);
+    }
+
+    /**
+     *  Schedule a trial and return allocated GPU indices.
+     *
+     *  If the trial cannot be scheduled for now, return `null`.
+     *  If the trial requires more then physical GPUs, throw error.
+     *
+     *  This scheduler does NOT monitor the trial's life span.
+     *  The caller must invokes `release()` or `releaseAll()` later.
+     *
+     *  `gpuNumber` can either be an integer or a float between 0 and 1.
+     *
+     *  The `restrictions` parameter may contain following options:
+     *
+     *  - onlyAcceptIndices:
+     *
+     *      Limit usable GPUs by index.
+     *      This is `gpuIndices` in experiment config.
+     *
+     *  - rejectActiveGpus:
+     *
+     *      Do not use GPUs with running processes.
+     *      This is reversed `useActiveGpu` in experiment config.
+     *
+     *  - rejectComputeActiveGpus:
+     *
+     *      Do not use GPUs with CUDA processes, but still use GPUs with graphics processes.
+     *      This is useful for desktop systems with graphical interface.
+     **/
+    public async schedule(
+            experimentId: string,
+            trialId: string,
+            gpuNumber: number,
+            restrictions?: ScheduleRestrictions): Promise<number[] | null> {
+
+        if (gpuNumber >= this.gpus.length) {
+            throw new Error(`GpuScheduler: Only have ${this.gpus.length} GPUs, requesting ${gpuNumber}`);
+        }
+
+        const gpus = this.sortGpus(restrictions ?? {});
+
+        if (gpuNumber < 1) {
```

Comment:
I think allowing `gpuNumber` to be a float is a bad design that will bring us troubles in future.

Maybe requestedGpuMemory is a better idea.

Reply:
We have `trial_number_per_gpu` currently. It will be translated to `gpu_number: gpu_number / trial_number_per_gpu` somewhere. So it is at least useful internally inside NNI manager. We can reject float `gpu_number` at python side validation.

Reply:
Seems ray is also doing so. I think it might work as well. https://docs.ray.io/en/latest/ray-core/tasks/using-ray-with-gpus.html#fractional-gpus